docs/building.md

```diff
@@ -0,0 +1,176 @@
+[one script](https://github.com/fobnail/fobnail-sdk/blob/main/run-fobnail-sdk.sh)
+from this repository is actually needed.
+
+Following steps were successfully performed on Ubuntu 22.04.1 LTS.
+
+## Fobnail Token firmware
+
+Start by installing prerequisites and adding current user to `docker` group:
+
+```shell
+sudo apt -y install git docker.io
+sudo usermod -aG docker $USER
+```
+
+Log out of the desktop session then log in again - this is required after making
+the change to user groups. You may confirm it by running `groups`.
+
+Clone the repository along with its submodules:
+
+```shell
+git clone https://github.com/fobnail/fobnail --recurse-submodules
+```
+
+Build process for Token firmware is by far the most complicated one. For that
+reason Fobnail SDK is strongly suggested. To use it, simply download script
+`run-fobnail-sdk.sh` either by cloning whole repository or just this one script
+and save it to a directory in your `PATH` environment variable:
+
+```shell
+sudo wget -O /usr/local/bin/run-fobnail-sdk.sh \
+  https://raw.githubusercontent.com/fobnail/fobnail-sdk/main/run-fobnail-sdk.sh
+sudo chmod +x /usr/local/bin/run-fobnail-sdk.sh
+```
+
+With the script installed, one can prepare for the build process itself.
+Firmware can be build for a physical Fobnail Token or simulated on PC, with
+different preparatory steps required. In both cases, first execution of
+`build.sh` downloads Docker container image, which can be both time- and
+bandwidth-consuming.
+
+#### Environment variables common for both targets
+
+Fobnail Token firmware is configured with environment variables passed to
+`build.sh`.
+
+Note that due to the way Docker mounts directories all files and directories
+pointed to by following variables must be located somewhere in `fobnail`
+directory.
+
+- `FOBNAIL_PO_ROOT` - **required** option, must point to valid PEM or DER file
+  with Platform Owner's root certificate. See [this document](/keys_and_certificates/#platform-owner-certificate-chain)
+  for description of PO certificate chain and instructions for building such.
+
+- `FOBNAIL_EK_ROOT_DIR` - points to directory with TPM root certificates.
+  Fobnail repository includes [such directory](https://github.com/fobnail/fobnail/tree/main/tpm_ek_roots)
+  which normally should be used as `FOBNAIL_EK_ROOT_DIR`, but you may change it
+  if certificates supplied in repository became outdated. Either this or
+  `FOBNAIL_EXTRA_EK_ROOT` (or both) **must be specified**, but build process
+  won't fail when neither is passed.
+
+- `FOBNAIL_EXTRA_EK_ROOT` - points to one specific TPM certificate. Useful when
+  testing with TPM emulator, or to limit Fobnail usage to one specific TPM
+  vendor. Either this or `FOBNAIL_EK_ROOT_DIR` (or both) **must be specified**,
+  but build process won't fail when neither is passed.
+
+#### Building and flashing firmware to Fobnail Token
+
+For hardware setup instructions see [Flashing preparation](flashing_preparation.md).
```

Comment:
> All 4 pieces of hardware are required for flashing nRF52840 dongle AFAIK. This document (flashing_preparations.md) doesn't apply for emulation. 

This is very surprising. I flashed OpenSK and previous Fobnail implementations without access to all those tools, I had just dongle, when those requirements changed?

> cargo-embed is included in Fobnail SDK, I'm working on document describing it currently.

Please link commit with that improvement, when it would be ready.

Reply:
> This is very surprising. I flashed OpenSK and previous Fobnail implementations without access to all those tools, I had just dongle, when those requirements changed?

Can you link to the instructions then? The way it is described here is the only way I know of, but I'm not saying there are no others. If there is a way of flashing it without external hardware it would definitely make it cheaper.

> Please link commit with that improvement, when it would be ready.

Done in a couple of commits, final result is in `docs/fobnail-sdk.md` on this branch.

Reply:
> Can you link to the instructions then? The way it is described here is the only way I know of, but I'm not saying there are no others. If there is a way of flashing it without external hardware it would definitely make it cheaper.

https://github.com/google/OpenSK/blob/stable/docs/install.md
https://github.com/google/OpenSK/blob/stable/docs/boards/nrf52840_dongle.md#flashing-using-dfu-preferred-method

Behind the scene [nrfutil](https://github.com/google/OpenSK/blob/stable/deploy.py#L706) is used I recall that in some cases I had to play little bit directly with `nrfutil` to make things work.

Reply:
@pietrushnic @krystian-hebel This is a bootloader feature, dongles come with bootloader pre-installed (which is part of Nordic SDK), we could flash Fobnail using DFU, but that would erase the bootloader (Fobnail firmware would have be prepared to work from another location in flash memory).

Reply:
If we want to have ability to flash the dongle without external hardware, it would have to have some kind of bootloader. You would still have to trust that bootloader's supply chain wasn't attacked.

I agree that current approach is both complicated and expensive, especially for personal use. Maybe it is possible to flash Fobnail firmware through OpenSK **once**, without the ability to change it later without external hardware, but we haven't tested such approach. We are open for suggestions, but so far we haven't been able to come up with anything easier that wouldn't impact security assumptions, especially those about firmware supply chain.

Reply:
> If we want to have ability to flash the dongle without external hardware, it would have to have some kind of bootloader. You would still have to trust that bootloader's supply chain wasn't attacked.

True and for those who need really high assurance cost of additional hardware to fully flash everything should not be problem. Key issue is at this stage of adoption where we would like to simplify as much as possible development and user testing cycle.

High assurance scenario would require also commercial approach and how potential customer can do things in scale, because most likely in such environment it would bring the most value. So key question here is if high assurance approach scale by any means?

> I agree that current approach is both complicated and expensive, especially for personal use. Maybe it is possible to flash Fobnail firmware through OpenSK once, without the ability to change it later without external hardware, but we haven't tested such approach. We are open for suggestions, but so far we haven't been able to come up with anything easier that wouldn't impact security assumptions, especially those about firmware supply chain.

Let's consider high assurance (full dev kit required) and low assurance (tester, developer, user which can trust 3mdeb). Low assurance approach would focus on extending users, testers and developers base, so it focus on simplicity and proving use cases, but does not provide security properties at high assurance level. Clear distinction have to be made. Please note OpenSK also explain it in their documentation by e.g.  [certificate considerations](https://github.com/google/OpenSK/blob/stable/docs/customization.md#Certificate-considerations) or [software personalization](https://github.com/google/OpenSK/blob/stable/docs/customization.md#software-personalization). We focusing on high assurance despite the fact that Fobnail is far from production, so maybe we should put emphasis on low assurance leaving note that high assurance can be achieved but it is more complex and until we provide business value proposition we will not invest much in high assurance direction.

Reply:
https://github.com/google/OpenSK/blob/stable/docs/install.md#programmers lists `nrfutil` in the last place. What's more, [this util was deprecated](https://github.com/NordicSemiconductor/pc-nrfutil#deprecation-notice) and [a closed-source tool](https://www.nordicsemi.com/Products/Development-tools/nrf-util) is available in its place. This doesn't bode well...

Reply:
Congratulations to Nordic. This is a terrible decision. So programming with binary blob is a bad idea.

I would be glad to resolve this discussion if a summary of it would be put in issue in which we will a aim for way more straightforward method of flashing, even if that means increasing potential attack surface. Essentially copy pasting with slight editing of my comment should be enough.

Reply:
https://github.com/fobnail/fobnail/issues/64